[Inductor CUTLASS backend] Step 2: CUDACodeCache

ipiszy opened the pull request:
This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108015
* #107931
* #107901
* __->__ #107847
* #107802



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

ipiszy committed:
CUDAcodecache

[ghstack-poisoned]

kadeng commented on torch/_inductor/codecache.py:
```diff
@@ -1186,6 +1187,254 @@ def load(cls, kernel_name, source_code):
 
 
+def _cuda_compiler() -> str:
+    return "nvcc"
```
Maybe we should use the same nvcc executable that is used and detected by torch's CMAKE build script, as that might not be the one that's on the path.

ipiszy replied:
Do you refer to `cmake/public/cuda.cmake`? From my understanding, this is written in CMake language and requires cmake to execute. I'm not familiar with CMake, and it doesn't look very trivial to do the integration. Since the cutlass integration is new, I'd like make it work first and then update the surrounding pieces, e.g. env setup, etc. For now I would just add a config in Inductor to let user specify nvcc path if they need.

jansel replied:
This won't work since CMAKE often runs on a different machine than pytorch gets `pip installed` on.

kadeng commented on torch/_inductor/codecache.py:
```diff
@@ -1186,6 +1187,254 @@ def load(cls, kernel_name, source_code):
+
+        if f_dlclose is not None:
+            f_dlclose.argtypes = [c_void_p]
+            f_dlclose(self.DLL._handle)
```
I think if this returns, we should set self.DLL = None

ipiszy replied:
All self.DLL access is guarded by self.is_open already.

kadeng commented on torch/_inductor/codecache.py:
```diff
@@ -1186,6 +1187,254 @@ def load(cls, kernel_name, source_code):
+        """
+
+        cuda_command = repr(
+            cuda_compile_command(["dummy_input"], "dummy_output", dst_file_ext)
```
Is this "dummy_input" intentional?

ipiszy replied:
Yes this is intentional. This is used to generate a dummy command for code hash purpose.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy reviewed (commented):
Thanks @kadeng !

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"

This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

aakhundov commented on torch/_inductor/codegen/cuda/cuda_env.py:
```diff
@@ -30,3 +30,14 @@ def get_cuda_version() -> Optional[str]:
         return None
+
+
+def nvcc_exist(nvcc_path: str = "nvcc") -> bool:
```
Nit: `Optional[str]`. mypy will be unhappy otherwise.

ipiszy replied:
Seems that it's okay.

aakhundov commented on torch/_inductor/codecache.py:
```diff
@@ -1186,6 +1187,263 @@ def load(cls, kernel_name, source_code):
+        extra_ldflags.append("-lcuda")
+        extra_ldflags.append("-lcudart")
+    else:
+        raise NotImplementedError("Unsupported env, failed to find cuda libs!")
```
If only Linux is supported at the moment, maybe let's be specific about it in the exception text?

aakhundov commented on torch/_inductor/codecache.py:
```diff
@@ -1186,6 +1187,263 @@ def load(cls, kernel_name, source_code):
+
+
+def _cuda_lib_options() -> List[str]:
+    from torch.utils import cpp_extension
```
Just in case: the other day I've found out that `cpp_extension` doesn't quite work internally (in my case: for inline compilation with `ninja`). Maybe for this particular use case it's fine, though.

ipiszy replied:
By internally, do you mean fbcode or something else? Could you help explain more?

aakhundov replied:
Yes, in `fbcode`. More context in this [comment](https://fburl.com/diff/b1xcu3e5). Maybe irrelevant for this PR, just a heads-up.

jansel commented on test/inductor/test_cudacodecache.py:
```diff
@@ -0,0 +1,89 @@
+
+__attribute__((__visibility__("default")))
+int saxpy(int n, float a, float *x, float *y) {
+  // Perform SAXPY
```
what is SAXPY?   Does it stand for something?

ipiszy replied:
Haha it's "Single-precision A*X Plus Y". This code is from https://developer.nvidia.com/blog/easy-introduction-cuda-c-and-c/.

ipiszy commented:
wrt how to handle cutlass header files / scripts, I'm having offline discussions with @atalman @jansel . Will report back.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"

This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented on test/inductor/test_cudacodecache.py:
```diff
@@ -0,0 +1,89 @@
+from torch._inductor.exc import CUDACompileError
+from torch.testing._internal.common_utils import TestCase as TorchTestCase
+
+_SOURCE_CODE = r"""
```
I feel it's okay since this file is only for `test_cudacodecache`, and the variable follows the underscore prefix convention to mark it as internal only.

ipiszy commented on torch/_inductor/codecache.py:
```diff
@@ -1186,6 +1187,263 @@ def load(cls, kernel_name, source_code):
+def _cutlass_include_paths() -> List[str]:
+    from torch.utils import cpp_extension
+
+    cutlass_path = os.path.join(cpp_extension._TORCH_PATH, "../third_party/cutlass")
```
Refined it a bit to fetch cutlass path from the default inductor config and allow users defining a custom cutlass path.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"

This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy reviewed (commented):
Thanks @jansel @aakhundov @kadeng  , addressed comments, ptal.

kadeng reviewed (approved):
looks good to me

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"

This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"

This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"

This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/cuda/cuda_env.py:
```diff
@@ -30,3 +30,14 @@ def get_cuda_version() -> Optional[str]:
         return None
+
+
+def nvcc_exist(nvcc_path: str = "nvcc") -> bool:
```
`functools.lru_cache` around this since `subprocess.call` is expensive.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"

This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"

This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"

This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"

This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented:
@pytorchbot label "topic: not user facing"

ipiszy commented:
@pytorch-bot merge

ipiszy commented:
@pytorchbot merge

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"

This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 2: CUDACodeCache"

This is the step 2 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]
